### PR TITLE
Support variable precision time timestamp

### PR DESCRIPTION
**Pull request by lpoulain**
## Description

Set X-Trino-Client-Capabilities to PARAMETRIC_DATETIME. Update the unit and integration tests accordingly.

## Non-technical explanation

When querying Trino, set the X-Trino-Client-Capabilities HTTP header to PARAMETRIC_DATETIME. This instructs Trino to return full-precision datetimes and times.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
* The client is now requesting PARAMETRIC_DATETIME, leading Trino to return full-precision datetimes and times. 
```

### CODE REVIEW

File: trino/client.py [hunk 1]
```diff
@@ -912,6 +919,7 @@ def round_to(self, precision: int) -> TemporalType:
             In case the supplied value exceeds the specified precision,
             the value needs to be rounded.
         """
+        precision = min(precision, 6)
```
Comment (mdesmet) [inline trino/client.py:922]: Please create a constant with a descriptive name for the max python temporal precision.

File: tests/integration/test_dbapi_integration.py [hunk 2]
```diff
@@ -311,7 +311,7 @@ def test_datetime_with_numeric_offset_time_zone_query_param(trino_connection):
     rows = cur.fetchall()
 
     assert rows[0][0] == params
-    assert cur.description[0][1] == "timestamp with time zone"
+    assert cur.description[0][1] == "timestamp(6) with time zone"
```
Comment (mdesmet) [inline tests/integration/test_dbapi_integration.py:314]: Also add similar tests for `time` and `time with time zone`. Seems the description is not tested there.

**Review by mdesmet — COMMENTED** [review]
LGTM % comments.

Also there is a merge commit, could you instead rebase on current master?

File: tests/integration/test_dbapi_integration.py (hunk 2 above)
Comment (lpoulain) [reply tests/integration/test_dbapi_integration.py:314]: Actually there are already `time` and `time with time zone` tests, e.g. `test_time_query_param`, `test_time_with_named_time_zone_query_param`

File: trino/client.py (hunk 1 above)
Comment (lpoulain) [reply trino/client.py]: Done

File: trino/client.py [hunk 3]
```diff
@@ -66,6 +66,7 @@
     PROXIES = {}
 
 _HEADER_EXTRA_CREDENTIAL_KEY_REGEX = re.compile(r'^\S[^\s=]*$')
+PYTHON_MAX_DECIMAL_PRECISION = 6
```
Comment (mdesmet) [inline trino/client.py:69]: Rename to `MAX_PYTHON_TEMPORAL_PRECISION_POWER`, move to line 77 and reuse in constant `MAX_PYTHON_TEMPORAL_PRECISION`.

File: tests/integration/test_dbapi_integration.py (hunk 2 above)
Comment (mdesmet) [reply tests/integration/test_dbapi_integration.py:314]: I meant there is no description test in those tests. Please add it to validate the type is correctly exposed.

File: trino/client.py [hunk 4]
```diff
@@ -747,7 +748,14 @@ def execute(self, additional_http_headers=None) -> TrinoResult:
         if self.cancelled:
             raise exceptions.TrinoUserError("Query has been cancelled", self.query_id)
 
-        response = self._request.post(self._sql, additional_http_headers)
+        if self._experimental_python_types:
```
Comment (mdesmet) [inline trino/client.py:751]: Actually you can always enable `PARAMETRIC_DATETIME`. This is also how the JDBC driver works. Please also adjust the commit message and PR description.

Comment (lpoulain) [reply trino/client.py:751]: Shouldn't this be part of another PR?

File: tests/integration/test_dbapi_integration.py (hunk 2 above)
Comment (lpoulain) [reply tests/integration/test_dbapi_integration.py:314]: Done

File: trino/client.py (hunk 3 above)
Comment (lpoulain) [reply trino/client.py]: Done

File: trino/client.py (hunk 4 above)
Comment (lpoulain) [reply trino/client.py]: Done

**Review by mdesmet — APPROVED** [review]

**Comment by mdesmet** [comment]
@lpoulain: Please squash the commits.

**Review by hovaesco — APPROVED** [review]

File: tests/integration/test_dbapi_integration.py [hunk 5]
```diff
@@ -461,6 +461,7 @@ def test_time_query_param(trino_connection):
     rows = cur.fetchall()
 
     assert rows[0][0] == params
+    assert cur.description[0][1] == "time(6)"
```
Comment (hashhar) [inline tests/integration/test_dbapi_integration.py:464]: same for `time with time zone`?

File: tests/integration/test_types_integration.py [hunk 6]
```diff
@@ -212,69 +212,74 @@ def test_time(trino_connection):
...
-        #     python=time(0, 0, 0))  # PARAMETRIC_DATETIME not enabled
-        # .add_field(
-        #     sql="TIME '23:59:59.99999'",
-        #     python=time(0, 0, 0))  # PARAMETRIC_DATETIME not enabled
-        # .add_field(
-        #     sql="TIME '23:59:59.999999'",
-        #     python=time(0, 0, 0))  # PARAMETRIC_DATETIME not enabled
-        # round down
         .add_field(
-            sql="TIME '00:00:00.000100'",
```
Comment (hashhar) [inline tests/integration/test_types_integration.py:227]: where did this case go?

File: tests/integration/test_types_integration.py [hunk 7]
```diff
@@ -212,69 +212,74 @@ def test_time(trino_connection):
...
-            python=time(0, 0, 0))
+            sql="TIME '23:59:59.9999'",
+            python=time(23, 59, 59, 999900))
         .add_field(
-            sql="TIME '00:00:00.0000001'",
-            python=time(0, 0, 0))  # PARAMETRIC_DATETIME not enabled
```
Comment (hashhar) [inline tests/integration/test_types_integration.py:231]: where did this go?

File: trino/client.py [hunk 8]
```diff
@@ -73,7 +73,8 @@
 POWERS_OF_TEN: Dict[int, Decimal] = {}
 for i in range(0, 13):
     POWERS_OF_TEN[i] = Decimal(10 ** i)
-MAX_PYTHON_TEMPORAL_PRECISION = POWERS_OF_TEN[6]
+MAX_PYTHON_TEMPORAL_PRECISION_POWER = 6
```
Comment (hashhar) [inline trino/client.py:76]:
```suggestion
MAX_PYTHON_TEMPORAL_PRECISION = 6
```

File: trino/client.py [hunk 9]
```diff
@@ -747,7 +748,11 @@ def execute(self, additional_http_headers=None) -> TrinoResult:
         if self.cancelled:
             raise exceptions.TrinoUserError("Query has been cancelled", self.query_id)
 
-        response = self._request.post(self._sql, additional_http_headers)
+        http_headers = {constants.HEADER_CLIENT_CAPABILITIES: 'PARAMETRIC_DATETIME'}
+        if additional_http_headers:
+            http_headers.update(additional_http_headers)
```
Comment (hashhar) [inline trino/client.py:753]: this must go into the `http_headers` method instead where we already do the request header setting logic.

**Review by hashhar — COMMENTED** [review]
LGTM % move header setting code to `http_headers` method

File: tests/integration/test_types_integration.py (hunk 6 above)
Comment (lpoulain) [reply tests/integration/test_types_integration.py:227]: Do we still need this considering there is no rounding involved here?

File: tests/integration/test_types_integration.py (hunk 7 above)
Comment (lpoulain) [reply tests/integration/test_types_integration.py:231]: Isn't this moved a few lines below (233)?

File: trino/client.py (hunk 9 above)
Comment (lpoulain) [reply trino/client.py]: Fixed

File: tests/integration/test_dbapi_integration.py (hunk 5 above)
Comment (lpoulain) [reply tests/integration/test_dbapi_integration.py:464]: Added similar assertion to `test_time_with_time_zone_negative_offset()` and `test_time_with_time_zone_positive_offset()`

File: tests/integration/test_types_integration.py (hunk 6 above)
Comment (hashhar) [reply tests/integration/test_types_integration.py:227]: Good catch, the rounding only happenned when parametric timestamp was not implemented. `00:00:00.000100` has `1` at 4th decimal digit and without parametric timestamp support it gets rounded to `00:00:00`.

File: tests/integration/test_types_integration.py (hunk 7 above)
Comment (hashhar) [reply tests/integration/test_types_integration.py:231]: Yeah, the diff on GitHub looked a bit wacky. I've re-reviewed locally in IDE and everything looks good.

**Review by hashhar — APPROVED** [review]
Please rebase/squash and resolve conflict. LGTM otherwise.

**Comment by hashhar** [comment]
rebased and reworded

**Review by hashhar — APPROVED** [review]

**Pull request merged**